sdk/spring/spring-cloud-azure-feature-management/src/test/java/com/azure/spring/cloud/feature/management/ValidationsTest.java

```diff
@@ -0,0 +1,152 @@
+        final String jsonString = Files.readString(testFile.toPath());
+        final CollectionType typeReference =
+            TypeFactory.defaultInstance().constructCollectionType(List.class, ValidationTestCase.class);
+        return objectMapper.readValue(jsonString, typeReference);
+    }
+
+    @SuppressWarnings("unchecked")
+    private LinkedHashMap<String, Object> readConfigurationFromFile(File sampleFile) throws IOException {
+        final String jsonString = Files.readString(sampleFile.toPath());
+        final LinkedHashMap<String, Object> configurations = objectMapper.readValue(jsonString, new TypeReference<>() {
+        });
+        final Object featureManagementSection = configurations.get("feature_management");
+        if (featureManagementSection.getClass().isAssignableFrom(LinkedHashMap.class)) {
+            return (LinkedHashMap<String, Object>) featureManagementSection;
+        }
+        return new LinkedHashMap<>();
+    }
+
+    @SuppressWarnings("unchecked")
+    private void runTestcases(File sampleFile, File testsFile) throws IOException {
+        // initialize feature manager
+        final FeatureManagementProperties managementProperties = new FeatureManagementProperties();
+        managementProperties.putAll(readConfigurationFromFile(sampleFile));
+        final FeatureManager featureManager = new FeatureManager(context, managementProperties, configProperties);
+
+        final List<ValidationTestCase> testCases = readTestcasesFromFile(testsFile);
+        System.out.println("Running test case from file: " + testsFile.getName());
+
+        for (int i = 0; i < testCases.size(); i++) {
+            System.out.println("Test case " + i + " : " + testCases.get(i).getDescription());
+            if (hasException(testCases.get(i))) {   // TODO. Currently we didn't throw the exception when parameter is invalid
+                assertNull(managementProperties.getOnOff().get(testCases.get(i).getFeatureFlagName()));
+            } else {
+                if (hasInput(testCases.get(i))) { // Set inputs
+                    final Object userObj = testCases.get(i).getInputs().get(inputsUser);
+                    final Object groupsObj = testCases.get(i).getInputs().get(inputsGroups);
+                    final String user = userObj != null ? userObj.toString() : null;
+                    final List<String> groups = groupsObj != null ? (List<String>) groupsObj : null;
+                    when(context.getBean(Mockito.contains("Targeting"))).thenReturn(new TargetingFilter(new TargetingFilterTestContextAccessor(user, groups)));
+                }
+
+                final Boolean result = featureManager.isEnabled(testCases.get(i).getFeatureFlagName());
+                assertEquals(result.toString(), testCases.get(i).getIsEnabled().getResult());
+            }
+        }
+    }
+
+    @Test
+    void validationsTest() throws IOException {
```

Comment:
Should we run each file as its own test? That's what I did in python. It makes it more controllable of which tests are running. Both methods have their pros and cons

Reply:
I also tried to have separate files and didn't mock the `FeatureManagementConfigProperties`. But I met 2 questions, not sure how to resolve.
```
@SpringBootTest(classes = { SpringBootTest.class })
@EnableConfigurationProperties({ FeatureManagementConfigProperties.class, FeatureManagementProperties.class })
@TestPropertySource(locations = { "classpath:/validations-tests/NoFilters.sample.json" }, factory= JsonPropertySourceFactory.class)
public class NoFiltersTest {

    private FeatureManager featureManager;

    @Test
    public void test() {
        featureManager.isEnabled("beta");
        // do something
    }

}
```

1. The `TestPropertySource` not work. The `FeatureManagementConfigProperties` was still initialized by "application.yaml" instead of "NoFilters.sample.json"
2. How to initialzied `featureManager`? Shall I still need to call `new FeatureManager()` to new instace? Or else

Reply:
Synced offline